versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/spi/AbstractDatabaseAdapter.java

```diff
@@ -825,6 +827,119 @@ public final CommitLogEntry fetchFromCommitLog(OP_CONTEXT ctx, Hash hash) {
 
   protected abstract CommitLogEntry doFetchFromCommitLog(OP_CONTEXT ctx, Hash hash);
 
+  @Override
+  public Stream<CommitLogEntry> scanAllCommitLogEntries() {
+    OP_CONTEXT ctx = borrowConnection();
+    return doScanAllCommitLogEntries(ctx)
+        .onClose(
+            () -> {
+              try {
+                ctx.close();
+              } catch (Exception e) {
+                throw new RuntimeException(e);
+              }
+            });
+  }
+
+  @Override
+  public HeadsAndCommonAncestors identifyAllHeadsAndCommonAncestors() {
+    Set<Hash> commits = new HashSet<>();
+    Set<Hash> heads = new HashSet<>();
+    Set<Hash> commonAncestors = new HashSet<>();
+
+    // Need to remember the time when the identification started, so that a follow-up
+    // identifyReferencedAndUnreferencedHeads() knows when it can stop scanning a named-reference's
+    // commit-log. identifyReferencedAndUnreferencedHeads() has to read up to the first commit
+    // _before_ this timestamp to not commit-IDs as "unreferenced".
+    //
+    // Note: keep in mind, that scanAllCommitLogEntries() returns all commits in a
+    // non-deterministic order. Example: if (at least) two commits are added to a branch while this
+    // function is running, the original HEAD of that branch could otherwise be returned as
+    // "unreferenced".
+    long scanStartedAtInMicros = this.commitTimeInMicros();
+
+    try (Stream<CommitLogEntry> scan = scanAllCommitLogEntries()) {
+      scan.forEach(
+          entry -> {
+            Hash parent = entry.getParents().get(0);
+            if (!commits.add(parent)) {
+              // If "parent" has already been added to the set of commits, then it must be a
+              // common ancestor.
+              commonAncestors.add(parent);
```

Comment:
> I'd rename it to "fork points" maybe?... just for the sake of clarity. In my mind a parent of a common ancestor is also a common ancestor thinking

We're using the term "common ancestor" elsewhere already (named-ref information, merges). Not sure if adding another term would be good.

> Might as well compute the number of branches stemming from this commit, I guess :)

That's what the other function `identifyReferencedAndUnreferencedHeads` does ;)

Reply:
In other code places the "common" ancestor is always in the context of two references, so the meaning is quite clear (at least for me :) ). Here, by contrast we're finding common ancestors even among unreferenced HEADs, so to me the meaning of "common" in this context is not as clear, because the entities that "share" a particular "common" commit themselves vary from case to case :thinking:

Reply:
As a side benefit we'd get the previous ref name for unreferenced HEADs, which might be relevant for access control and historical investigation (by end users).

Reply:
Renamed to "fork point"